### PR TITLE
refactor(components)!: [select-v2] value-key is used for object select

### LINKED ISSUE #6502: [Bug Report] [Component] [select-v2] SelectV2在options的value是一个对象时，判断option是否选中的逻辑有问题
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.31`
- Element Plus Version: `2.0.3`
- Browser / OS: `Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) biz-frame-generator/1.0.0 Chrome/98.0.4758.102 Electron/17.1.0 Safari/537.36`
- Build Tool: `Vue CLI`

### Related Component

- `el-select-v2`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcblxuY29uc3Qgb3B0aW9ucyA9IFt7XG4gIHZhbHVlOiB7XG4gICAgY29kZTogJzEnLFxuICAgIG5hbWU6ICd4J1xuICB9LFxuICBsYWJlbDogJ3gnXG59LCB7XG4gIHZhbHVlOiB7XG4gICAgY29kZTogJzInLFxuICAgIG5hbWU6ICd5J1xuICB9LFxuICBsYWJlbDogJ3knXG59LCB7XG4gIHZhbHVlOiB7XG4gICAgY29kZTogJzMnLFxuICAgIG5hbWU6ICd6J1xuICB9LFxuICBsYWJlbDogJ3onXG59XTtcblxuY29uc3Qgc2VsZWN0ZWQgPSByZWYoW10pO1xuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPGVsLXNlbGVjdC12MiB2LW1vZGVsPVwic2VsZWN0ZWRcIiA6b3B0aW9ucz1cIm9wdGlvbnNcIiBtdWx0aXBsZSBzdHlsZT1cIndpZHRoOiAyMDBweFwiLz5cbjwvdGVtcGxhdGU+IiwiaW1wb3J0LW1hcC5qc29uIjoie1wiaW1wb3J0c1wiOnt9fSJ9)

### Steps to reproduce

功能上希望选择多个对象，SelectV2似乎不支持直接在v-model中返回options中的对象，也不支持指定label的字段名，因此在每个对象外又套了一层，把对象放在value上，把显示名放在label上

### What is Expected?

希望在选中后，能在v-model中返回value的数组

### What is actually happening?

选中一个，下拉列表中所有都显示选中，似乎所有选项都被当成了一个选项

### Additional comments

查看了一下源码，在value是对象时，存在两次取valueKey的情况。

useSelect.ts中，onSelect函数405行在getValueIndex入参时已经取了一次valueKey，在getValueIndex内部如果value是对象，又会取一次valueKey，在select-dropdown.vue的isItemSelected中也有同样的问题，只不过第一次是写死取value，这个是否有问题

### LINKED ISSUE #7140: [Bug Report] [Component] [select-v2] el-select-v2 v-model使用object时候，value-key无法使用
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.26`
- Element Plus Version: `2.1.9`
- Browser / OS: `Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.60 Safari/537.36 Edg/100.0.1185.29`
- Build Tool: `Vite`

### Related Component

- `el-select-v2`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZWwtc2VsZWN0LXYyXG4gICAgdi1tb2RlbD1cInZhbHVlXCJcbiAgICA6b3B0aW9ucz1cIm9wdGlvbnNcIlxuICAgIHBsYWNlaG9sZGVyPVwiUGxlYXNlIHNlbGVjdFwiXG4gICAgc2l6ZT1cImxhcmdlXCJcbiAgICB2YWx1ZS1rZXk9XCJpZFwiXG4gIC8+XG48L3RlbXBsYXRlPlxuXG48c2NyaXB0IGxhbmc9XCJ0c1wiIHNldHVwPlxuaW1wb3J0IHsgcmVmIH0gZnJvbSAndnVlJ1xuY29uc3QgaW5pdGlhbHMgPSBbe1xuICBpZDogMSxcbiAgbmFtZTogJ2EnXG59LHtcbiAgaWQ6IDIsXG4gIG5hbWU6ICdiJ1xufSx7XG4gIGlkOjMsXG4gIG5hbWU6ICdjJ1xufV1cblxuY29uc3QgdmFsdWUgPSByZWYoJycpXG5jb25zdCBvcHRpb25zID0gaW5pdGlhbHMubWFwKChpdGVtLCBpZHgpID0+ICh7XG4gIHZhbHVlOiBpdGVtLFxuICBsYWJlbDogaXRlbS5uYW1lLFxufSkpXG48L3NjcmlwdD5cblxuPHN0eWxlIHNjb3BlZD5cbi5leGFtcGxlLXNob3djYXNlIC5lbC1zZWxlY3QtdjIge1xuICBtYXJnaW4tcmlnaHQ6IDIwcHg7XG59XG48L3N0eWxlPlxuIiwiaW1wb3J0LW1hcC5qc29uIjoie1wiaW1wb3J0c1wiOnt9fSJ9)

### Steps to reproduce

如链接所示。

### What is Expected?

选择a时候，value应该为{id:1, name: 'a'},
选择b时候，value应该为{id:2, name: 'b'},
选择c时候，value应该为{id:3, name: 'c'}

### What is actually happening?

报错：Cannot read properties of undefined (reading 'toString')

### Additional comments

_(empty)_

### LINKED ISSUE #8369: [Bug Report] [Component] [select-v2] el-select-v2 v-model绑定的值始终string
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.37`
- Element Plus Version: `2.2.6`
- Browser / OS: `UserAgent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.0.0 Safari/537.36`
- Build Tool: `Vite`

### Related Component

- `el-select-v2`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiwgdmVyc2lvbiBhcyB2dWVWZXJzaW9uIH0gZnJvbSAndnVlJ1xuaW1wb3J0IHsgdmVyc2lvbiBhcyBFcFZlcnNpb24gfSBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBFbGVtZW50UGx1cyB9IGZyb20gJ0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlJ1xuXG5jb25zdCB2YWx1ZSA9IHJlZignJylcblxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgIDxlbC1zZWxlY3QtdjJcbiAgICB2LW1vZGVsPVwidmFsdWVcIlxuICAgIHZhbHVlLWtleT1cInZhbHVlLmlkXCJcbiAgICA6b3B0aW9ucz1cIltcbiAgICAgIHtcbiAgICAgICBsYWJlbDogJ+a1i+ivlTEnLFxuICAgICAgIHZhbHVlOiB7XG4gICAgICAgICAgaWQ6IDFcbiAgICAgICB9XG4gICAgICB9LFxuICAgICB7XG4gICAgICAgbGFiZWw6ICfmtYvor5UyJyxcbiAgICAgICB2YWx1ZToge1xuICAgICAgICAgIGlkOiAyXG4gICAgICAgfVxuICAgICB9LFxuICAgIF1cIlxuICAgIHBsYWNlaG9sZGVyPVwiUGxlYXNlIHNlbGVjdFwiXG4gIC8+XG4gXHQ8cD5cbiAgICDpgInkuK3nmoTlgLx7e3ZhbHVlfX1cbiAgPC9wPlxuXHRcbiAgXG4gIDxwPlxuICAgIDxlbC1pY29uIGNvbG9yPVwidmFyKC0tZWwtY29sb3ItcHJpbWFyeSlcIj48RWxlbWVudFBsdXMgLz48L2VsLWljb24+XG4gICAgRWxlbWVudCBQbHVzIHt7IEVwVmVyc2lvbiB9fSArIFZ1ZSB7eyB2dWVWZXJzaW9uIH19XG4gIDwvcD5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7fVxufSIsIl9vIjp7fX0=)

### Steps to reproduce

 <el-select-v2
    v-model="value"
    value-key="value.id"
    :options="[
      {
       label: ‘测试1’,
       value: {
          id: 1
       }
      },
     {
       label: ‘测试2’,
       value: {
          id: 2
       }
     },
    ]"
    placeholder="Please select"
  />

### What is Expected?

选中后 v-model 绑定的值为预期的对象

### What is actually happening?

v-model 绑定的值为 测试1

### Additional comments

_(empty)_

### LINKED ISSUE #9628: [Component] [select-v2] [Component] [Select V2 ] el-select-v2 multiple options为对象数组时候数据回显无法显示label
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.33`
- Element Plus Version: `2.2.16`
- Browser / OS: `macOS BigS Sur 11.4`
- Build Tool: `Webpack`

### Related Component

- `el-select-v2`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcblxuY29uc3Qgb3B0aW9ucyA9IFt7XG4gIGNvZGU6IHtcbiAgICBjb2RlOiAnMScsXG4gICAgbmFtZTogJ3gnXG4gIH0sXG4gIGxhYmVsOiAneCdcbn0sIHtcbiAgY29kZToge1xuICAgIGNvZGU6ICcyJyxcbiAgICBuYW1lOiAneSdcbiAgfSxcbiAgbGFiZWw6ICd5J1xufSwge1xuICBjb2RlOiB7XG4gICAgY29kZTogJzMnLFxuICAgIG5hbWU6ICd6J1xuICB9LFxuICBsYWJlbDogJ3onXG59XTtcblxuY29uc3Qgc2VsZWN0ZWQgPSByZWYoW3tjb2RlOicyJyxcbiAgbmFtZToneSd9XSk7XG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICB7eyBzZWxlY3RlZCB9fVxuICA8ZWwtc2VsZWN0LXYyIHYtbW9kZWw9XCJzZWxlY3RlZFwiIDpvcHRpb25zPVwib3B0aW9uc1wiIG11bHRpcGxlIHN0eWxlPVwid2lkdGg6IDIwMHB4XCIgdmFsdWUta2V5PVwiY29kZVwiLz5cbjwvdGVtcGxhdGU+IiwiaW1wb3J0X21hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge31cbn0iLCJfbyI6e319)

### Steps to reproduce

el-select-v2 v-model绑定值赋值为对象数组，对象格式和options中的对象格式一致

### What is Expected?

el-select-v2下拉选项选中且正确显示label

### What is actually happening?

el-select-v2下拉选项命中但label没有显示

### Additional comments

_(empty)_

### PR DESCRIPTION
**Pull request by humorHan**
close #9628, close #8369, close #7140, close #6502, close #8374

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

Currently the `value-key` is used to select both value in `options` ​​and value in `modelValue`. But they are always preceded by a difference `value`, leading to inability to select objects or various other errors.

```
const options = [
  {
    value: {
      id: '1',
      name: 'x',
    },
    label: 'x',
  }
]
const modelValue = {  id: '1',  name: 'x'  }
```

value-key should be consistent with the Select component, only used to select objects.

### TODO

Added attribute(`props: { label: 'label', value: 'value', options: 'options' }`) to configure aliases for data in `options`

- [before](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcblxuY29uc3Qgc2VsZWN0ZWQgPSByZWYoe1xuICBpZDogJzInLFxuICBuYW1lOiAneScsXG59KVxuY29uc3QgbXVsdGlwbGUgPSByZWYoW1xuICB7XG4gICAgaWQ6ICcyJyxcbiAgICBuYW1lOiAneScsXG4gIH0sXG5dKVxuXG5jb25zdCBvcHRpb25zID0gW1xuICB7XG4gICAgdmFsdWU6IHtcbiAgICAgIGlkOiAnMScsXG4gICAgICBuYW1lOiAneCcsXG4gICAgfSxcbiAgICBsYWJlbDogJ3gnLFxuICB9LFxuICB7XG4gICAgdmFsdWU6IHtcbiAgICAgIGlkOiAnMicsXG4gICAgICBuYW1lOiAneScsXG4gICAgfSxcbiAgICBsYWJlbDogJ3knLFxuICB9LFxuICB7XG4gICAgdmFsdWU6IHtcbiAgICAgIGlkOiAnMycsXG4gICAgICBuYW1lOiAneicsXG4gICAgfSxcbiAgICBsYWJlbDogJ3onLFxuICB9LFxuXVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPGgxPlVuYWJsZSB0byBzZWxlY3Qgb2JqZWN0PC9oMT5cbiAgPGgyPm11bHRpcGxlOiBmYWxzZTwvaDI+XG4gIDxwPnt7IHNlbGVjdGVkIH19PC9wPlxuICA8ZGl2PlxuICAgIDxzcGFuIGNsYXNzPVwibGFiZWxcIj5zZWxlY3QtdjI8L3NwYW4+XG4gICAgPGVsLXNlbGVjdC12MlxuICAgICAgdi1tb2RlbD1cInNlbGVjdGVkXCJcbiAgICAgIDpvcHRpb25zPVwib3B0aW9uc1wiXG4gICAgICB2YWx1ZS1rZXk9XCJpZFwiXG4gICAgICBjbGVhcmFibGVcbiAgICAvPlxuICA8L2Rpdj5cbiAgPGRpdj5cbiAgICA8c3BhbiBjbGFzcz1cImxhYmVsXCI+c2VsZWN0PC9zcGFuPlxuICAgIDxlbC1zZWxlY3Qgdi1tb2RlbD1cInNlbGVjdGVkXCIgdmFsdWUta2V5PVwiaWRcIiBkaXNhYmxlZD5cbiAgICAgIDxlbC1vcHRpb25cbiAgICAgICAgdi1mb3I9XCJpdGVtIGluIG9wdGlvbnNcIlxuICAgICAgICA6a2V5PVwiaXRlbS5sYWJlbFwiXG4gICAgICAgIDpsYWJlbD1cIml0ZW0ubGFiZWxcIlxuICAgICAgICA6dmFsdWU9XCJpdGVtLnZhbHVlXCJcbiAgICAgIC8+XG4gICAgPC9lbC1zZWxlY3Q+XG4gIDwvZGl2PlxuICA8aDI+bXVsdGlwbGU6IHRydWU8L2gyPlxuICA8cD57eyBtdWx0aXBsZSB9fTwvcD5cbiAgPGRpdj5cbiAgICA8c3BhbiBjbGFzcz1cImxhYmVsXCI+c2VsZWN0LXYyPC9zcGFuPlxuICAgIDxlbC1zZWxlY3QtdjJcbiAgICAgIHYtbW9kZWw9XCJtdWx0aXBsZVwiXG4gICAgICA6b3B0aW9ucz1cIm9wdGlvbnNcIlxuICAgICAgbXVsdGlwbGVcbiAgICAgIHZhbHVlLWtleT1cImlkXCJcbiAgICAgIGNsZWFyYWJsZVxuICAgICAgc3R5bGU9XCJ3aWR0aDogMjAwcHhcIlxuICAgIC8+XG4gIDwvZGl2PlxuICA8ZGl2PlxuICAgIDxzcGFuIGNsYXNzPVwibGFiZWxcIj5zZWxlY3Q8L3NwYW4+XG4gICAgPGVsLXNlbGVjdCB2LW1vZGVsPVwibXVsdGlwbGVcIiB2YWx1ZS1rZXk9XCJpZFwiIG11bHRpcGxlIGRpc2FibGVkPlxuICAgICAgPGVsLW9wdGlvblxuICAgICAgICB2LWZvcj1cIml0ZW0gaW4gb3B0aW9uc1wiXG4gICAgICAgIDprZXk9XCJpdGVtLmxhYmVsXCJcbiAgICAgICAgOmxhYmVsPVwiaXRlbS5sYWJlbFwiXG4gICAgICAgIDp2YWx1ZT1cIml0ZW0udmFsdWVcIlxuICAgICAgLz5cbiAgICA8L2VsLXNlbGVjdD5cbiAgPC9kaXY+XG48L3RlbXBsYXRlPlxuXG48c3R5bGU+XG4ubGFiZWwge1xuICBkaXNwbGF5OiBpbmxpbmUtYmxvY2s7XG4gIG1hcmdpbi1yaWdodDogMTZweDtcbiAgd2lkdGg6IDEwMHB4O1xuICB0ZXh0LWFsaWduOiByaWdodDtcbn1cbjwvc3R5bGU+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7fVxufSIsIl9vIjp7fX0=)

- [after](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcblxuY29uc3Qgc2VsZWN0ZWQgPSByZWYoe1xuICBpZDogJzInLFxuICBuYW1lOiAneScsXG59KVxuY29uc3QgbXVsdGlwbGUgPSByZWYoW1xuICB7XG4gICAgaWQ6ICcyJyxcbiAgICBuYW1lOiAneScsXG4gIH0sXG5dKVxuXG5jb25zdCBvcHRpb25zID0gW1xuICB7XG4gICAgdmFsdWU6IHtcbiAgICAgIGlkOiAnMScsXG4gICAgICBuYW1lOiAneCcsXG4gICAgfSxcbiAgICBsYWJlbDogJ3gnLFxuICB9LFxuICB7XG4gICAgdmFsdWU6IHtcbiAgICAgIGlkOiAnMicsXG4gICAgICBuYW1lOiAneScsXG4gICAgfSxcbiAgICBsYWJlbDogJ3knLFxuICB9LFxuICB7XG4gICAgdmFsdWU6IHtcbiAgICAgIGlkOiAnMycsXG4gICAgICBuYW1lOiAneicsXG4gICAgfSxcbiAgICBsYWJlbDogJ3onLFxuICB9LFxuXVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPGgxPlVuYWJsZSB0byBzZWxlY3Qgb2JqZWN0PC9oMT5cbiAgPGgyPm11bHRpcGxlOiBmYWxzZTwvaDI+XG4gIDxwPnt7IHNlbGVjdGVkIH19PC9wPlxuICA8ZGl2PlxuICAgIDxzcGFuIGNsYXNzPVwibGFiZWxcIj5zZWxlY3QtdjI8L3NwYW4+XG4gICAgPGVsLXNlbGVjdC12MlxuICAgICAgdi1tb2RlbD1cInNlbGVjdGVkXCJcbiAgICAgIDpvcHRpb25zPVwib3B0aW9uc1wiXG4gICAgICB2YWx1ZS1rZXk9XCJpZFwiXG4gICAgICBjbGVhcmFibGVcbiAgICAvPlxuICA8L2Rpdj5cbiAgPGRpdj5cbiAgICA8c3BhbiBjbGFzcz1cImxhYmVsXCI+c2VsZWN0PC9zcGFuPlxuICAgIDxlbC1zZWxlY3Qgdi1tb2RlbD1cInNlbGVjdGVkXCIgdmFsdWUta2V5PVwiaWRcIiBkaXNhYmxlZD5cbiAgICAgIDxlbC1vcHRpb25cbiAgICAgICAgdi1mb3I9XCJpdGVtIGluIG9wdGlvbnNcIlxuICAgICAgICA6a2V5PVwiaXRlbS5sYWJlbFwiXG4gICAgICAgIDpsYWJlbD1cIml0ZW0ubGFiZWxcIlxuICAgICAgICA6dmFsdWU9XCJpdGVtLnZhbHVlXCJcbiAgICAgIC8+XG4gICAgPC9lbC1zZWxlY3Q+XG4gIDwvZGl2PlxuICA8aDI+bXVsdGlwbGU6IHRydWU8L2gyPlxuICA8cD57eyBtdWx0aXBsZSB9fTwvcD5cbiAgPGRpdj5cbiAgICA8c3BhbiBjbGFzcz1cImxhYmVsXCI+c2VsZWN0LXYyPC9zcGFuPlxuICAgIDxlbC1zZWxlY3QtdjJcbiAgICAgIHYtbW9kZWw9XCJtdWx0aXBsZVwiXG4gICAgICA6b3B0aW9ucz1cIm9wdGlvbnNcIlxuICAgICAgbXVsdGlwbGVcbiAgICAgIHZhbHVlLWtleT1cImlkXCJcbiAgICAgIGNsZWFyYWJsZVxuICAgICAgc3R5bGU9XCJ3aWR0aDogMjAwcHhcIlxuICAgIC8+XG4gIDwvZGl2PlxuICA8ZGl2PlxuICAgIDxzcGFuIGNsYXNzPVwibGFiZWxcIj5zZWxlY3Q8L3NwYW4+XG4gICAgPGVsLXNlbGVjdCB2LW1vZGVsPVwibXVsdGlwbGVcIiB2YWx1ZS1rZXk9XCJpZFwiIG11bHRpcGxlIGRpc2FibGVkPlxuICAgICAgPGVsLW9wdGlvblxuICAgICAgICB2LWZvcj1cIml0ZW0gaW4gb3B0aW9uc1wiXG4gICAgICAgIDprZXk9XCJpdGVtLmxhYmVsXCJcbiAgICAgICAgOmxhYmVsPVwiaXRlbS5sYWJlbFwiXG4gICAgICAgIDp2YWx1ZT1cIml0ZW0udmFsdWVcIlxuICAgICAgLz5cbiAgICA8L2VsLXNlbGVjdD5cbiAgPC9kaXY+XG48L3RlbXBsYXRlPlxuXG48c3R5bGU+XG4ubGFiZWwge1xuICBkaXNwbGF5OiBpbmxpbmUtYmxvY2s7XG4gIG1hcmdpbi1yaWdodDogMTZweDtcbiAgd2lkdGg6IDEwMHB4O1xuICB0ZXh0LWFsaWduOiByaWdodDtcbn1cbjwvc3R5bGU+XG4iLCJQbGF5Z3JvdW5kTWFpbi52dWUiOiI8c2NyaXB0IHNldHVwPlxuaW1wb3J0IEFwcCBmcm9tICcuL0FwcC52dWUnXG5pbXBvcnQgeyBzZXR1cEVsZW1lbnRQbHVzIH0gZnJvbSAnLi9lbGVtZW50LXBsdXMuanMnXG5zZXR1cEVsZW1lbnRQbHVzKClcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxBcHAgLz5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJlbGVtZW50LXBsdXNcIjogXCJodHRwczovL3ByZXZpZXctOTYzOC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2luZGV4LmZ1bGwubWluLm1qc1wiLFxuICAgIFwiZWxlbWVudC1wbHVzL1wiOiBcInVuc3VwcG9ydGVkXCJcbiAgfVxufSIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHtcbiAgICBcInZ1ZVwiOiBcImh0dHBzOi8vdW5wa2cuY29tL0B2dWUvcnVudGltZS1kb21AbGF0ZXN0L2Rpc3QvcnVudGltZS1kb20uZXNtLWJyb3dzZXIuanNcIixcbiAgICBcIkB2dWUvc2hhcmVkXCI6IFwiaHR0cHM6Ly91bnBrZy5jb20vQHZ1ZS9zaGFyZWRAbGF0ZXN0L2Rpc3Qvc2hhcmVkLmVzbS1idW5kbGVyLmpzXCIsXG4gICAgXCJlbGVtZW50LXBsdXNcIjogXCJodHRwczovL3ByZXZpZXctOTYzOC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2luZGV4LmZ1bGwubWluLm1qc1wiLFxuICAgIFwiZWxlbWVudC1wbHVzL1wiOiBcInVuc3VwcG9ydGVkXCIsXG4gICAgXCJAZWxlbWVudC1wbHVzL2ljb25zLXZ1ZVwiOiBcImh0dHBzOi8vdW5wa2cuY29tL0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlQDIvZGlzdC9pbmRleC5taW4uanNcIlxuICB9LFxuICBcInNjb3Blc1wiOiB7fVxufSIsImVsZW1lbnQtcGx1cy5qcyI6ImltcG9ydCB7IGdldEN1cnJlbnRJbnN0YW5jZSB9IGZyb20gJ3Z1ZSdcbmltcG9ydCBFbGVtZW50UGx1cyBmcm9tICdlbGVtZW50LXBsdXMnXG5cbmxldCBpbnN0YWxsZWQgPSBmYWxzZVxuYXdhaXQgbG9hZFN0eWxlKClcblxuZXhwb3J0IGZ1bmN0aW9uIHNldHVwRWxlbWVudFBsdXMoKSB7XG4gIGlmIChpbnN0YWxsZWQpIHJldHVyblxuICBjb25zdCBpbnN0YW5jZSA9IGdldEN1cnJlbnRJbnN0YW5jZSgpXG4gIGluc3RhbmNlLmFwcENvbnRleHQuYXBwLnVzZShFbGVtZW50UGx1cylcbiAgaW5zdGFsbGVkID0gdHJ1ZVxufVxuXG5leHBvcnQgZnVuY3Rpb24gbG9hZFN0eWxlKCkge1xuICByZXR1cm4gbmV3IFByb21pc2UoKHJlc29sdmUsIHJlamVjdCkgPT4ge1xuICAgIGNvbnN0IGxpbmsgPSBkb2N1bWVudC5jcmVhdGVFbGVtZW50KCdsaW5rJylcbiAgICBsaW5rLnJlbCA9ICdzdHlsZXNoZWV0J1xuICAgIGxpbmsuaHJlZiA9ICdodHRwczovL3ByZXZpZXctOTYzOC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2luZGV4LmNzcydcbiAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2xvYWQnLCByZXNvbHZlKVxuICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignZXJyb3InLCByZWplY3QpXG4gICAgZG9jdW1lbnQuYm9keS5hcHBlbmQobGluaylcbiAgfSlcbn0iLCJfbyI6eyJzaG93SGlkZGVuIjp0cnVlLCJzdHlsZVNvdXJjZSI6Imh0dHBzOi8vcHJldmlldy05NjM4LWVsZW1lbnQtcGx1cy5zdXJnZS5zaC9idW5kbGUvaW5kZXguY3NzIn19)

### CODE REVIEW

**Comment by humorHan** [comment]
@tolking 辛苦check下~

**Comment by humorHan** [comment]
> At

已修复单测问题

**Comment by humorHan** [comment]
> @humorHan 需要先解决下单元测试的问题哦。

已修复~

**Comment by humorHan** [comment]
> @humorHan 需要先解决下单元测试的问题哦。

@chenxch 已修复

File: packages/components/select-v2/src/useSelect.ts [hunk 1]
```diff
@@ -649,9 +649,13 @@ const useSelect = (props: ExtractPropTypes<typeof SelectProps>, emit) => {
...
         ;(props.modelValue as Array<any>).forEach((selected) => {
-          const itemIndex = filteredOptions.value.findIndex(
-            (option) => getValueKey(option) === selected
-          )
+          const itemIndex = filteredOptions.value.findIndex((option) => {
+            const _value = getValueKey(option)
+            return (
+              (isObject(_value) ? _value?.[props.valueKey] : _value) ===
+              getValueKey(selected)
+            )
```
Comment (tolking) [inline packages/components/select-v2/src/useSelect.ts:657]: 这样似乎不能够解决问题，如果你测试用例的两处 `value` 不一样问题依然存在

应该先使用 toRawType 判断 selected 的类型，如果选择的是对象则使用 isEqual 进行判断

Comment (tolking) [reply packages/components/select-v2/src/useSelect.ts:657]: 我想应该写一个共公的方法处理这个错误，这个错误在单选和多选都会存在，还包括 select-dropdown 中的 isItemSelected 判断

Comment (humorHan) [reply packages/components/select-v2/src/useSelect.ts:657]: > 这样似乎不能够解决问题，如果你测试用例的两处 `value` 不一样问题依然存在
> 
> 应该先使用 toRawType 判断 selected 的类型，如果选择的是对象则使用 isEqual 进行判断

如果是对象使用 isEqual 进行判断好像也不太满足预期，比如`[{value: 1, enName: 'san zhang', label: '张三'}]`， selected为{value:1}，这样并不能命中，感觉还是需要根据value-key找到简单类型字段进行判断。

另外再改动之前还想确认个事情，select-v2里的value字段是写死的，那么数据层嵌套也应该是写死吧？比如`[{value: {value: 'xxx'}, label: '张三'}]`，如此，那么`
(isObject(_value) ? _value?.[props.valueKey] : _value) ===
              getValueKey(selected)
`  是成立的。

Comment (tolking) [reply packages/components/select-v2/src/useSelect.ts:657]: 我记起来，现在内部 label 和 value 等是写死的。或许应该按照 Select 的逻辑 value-key 仅作用于 value 内部的对象，不然选择对象时，无法进行判断。

或许应该取`option.value` 进行判断 `(option) => getValueKey(option.value) === getValueKey(selected)`

Comment (tolking) [reply packages/components/select-v2/src/useSelect.ts:657]: 代码内部很多地方将两者混在一起了，估计有很多地方需要调整

Comment (humorHan) [reply packages/components/select-v2/src/useSelect.ts]: > 我记起来，现在内部 label 和 value 等是写死的。或许应该按照 Select 的逻辑 value-key 仅作用于 value 内部的对象，不然选择对象时，无法进行判断。
> 
> 或许应该取`option.value` 进行判断 `(option) => getValueKey(opti`

OK，那我也写死value吧...

Comment (humorHan) [reply packages/components/select-v2/src/useSelect.ts]: > 代码内部很多地方将两者混在一起了，估计有很多地方需要调整

只改了一处，发现两处类似的，但是因为入参不同不需要修改，我提交pr，辛苦check下

**Comment by tolking** [comment]
> @tolking 你好，请问当前pr还有需要完善的吗？以及预计发布时间？

这项改动有一定的破坏性，请耐心等待审核。如果不存在冲突文件，可以不用合并 dev 分支

**Comment by yigeqiji** [comment]
Is there a solution yet, the latest version(2.2.16) still has this problem?

**Comment by humorHan** [comment]
> > @tolking 你好，请问当前pr还有需要完善的吗？以及预计发布时间？
> 
> 这项改动有一定的破坏性，请耐心等待审核。如果不存在冲突文件，可以不用合并 dev 分支

好的，辛苦了~   
ps：本次上线已经绕过了这个问题，下个月版本迭代可能还会用到，目前是只是持续跟进一下

**Review by holazz — APPROVED** [review]

**Comment by sxzz** [comment]
@holazz Conflicts resolved.

**Comment by holazz** [comment]
> @holazz Conflicts resolved.

Okay, let's merge it.

**Comment by humorHan** [comment]
请问这个pr已经merge发版了吗？在changelog里一直没有看到该pr @tolking

**Comment by tolking** [comment]
@humorHan 要等 2.3

**Comment by mystylemylife** [comment]
我什么我最用了最新版本,该问题还是存在,绑定对象不管用

**Pull request merged**